### PR TITLE
plugin/hpsa: Use LocalDisk API to retrieve volume path

### PR DESCRIPTION
**Pull request by joehandzik**
When hpssacli fails to report a SCSI device path for a given volume, try
to get that path from LocalDisk. If LocalDisk also fails, avoid using
the path entirely.

Signed-Off-By: Joe Handzik <joseph.t.handzik@hpe.com>

### CODE REVIEW

**Comment by joehandzik** [comment]
@cvedel, can you give this a shot and confirm that it fixes your problem?

File: plugin/hpsa/hpsa.py [hunk 1]
```diff
@@ -573,18 +573,28 @@ def _hp_ld_to_lsm_vol(hp_ld, pool_id, sys_id, ctrl_num, array_num,
...
+            # No document or command output indicate block size
+            # of volume. So we try to read from linux kernel, if failed
+            # try 512 and roughly calculate the sector count.
+            try:
+                block_size = attributes.asint("queue/logical_block_size")
+                num_of_blocks = attributes.asint("size")
+            except (KeyError, UnicodeDecodeError, ValueError):
+                block_size = 512
+                num_of_blocks = int(
+                    _hp_size_to_lsm(hp_ld['Size']) / block_size)
```
Comment (cathay4t) [inline plugin/hpsa/hpsa.py]: `block_size` and `num_of_blocks` might be undefined when `hp_ld_path` is empty which might be possible when certain LV is not exposed to OS.

Comment (joehandzik) [inline plugin/hpsa/hpsa.py]: Good point. Any suggestions for clean ways to improve this? I'm thinking I should move the assignment of 'attributes' inside the try block, and figure out the right way to default block_size, num_of_blocks, and the values required for vol_name to sensible values.

Comment (cathay4t) [inline plugin/hpsa/hpsa.py]: Please try this patch: https://github.com/cathay4t/libstoragemgmt/commit/4a179265d748c03288c68d2b309095e4cd929cf9 
```patch
From 4a179265d748c03288c68d2b309095e4cd929cf9 Mon Sep 17 00:00:00 2001
From: Gris Ge <fge@redhat.com>
Date: Wed, 20 Jul 2016 09:33:41 +0800
Subject: [PATCH] HPSA Plugin: Use LocalDisk API to retrieve volume path

Fixing issue https://github.com/libstorage/libstoragemgmt/issues/167

Signed-off-by: Gris Ge <fge@redhat.com>
---
 plugin/hpsa/hpsa.py | 25 ++++++++++++++++---------
 1 file changed, 16 insertions(+), 9 deletions(-)

diff --git a/plugin/hpsa/hpsa.py b/plugin/hpsa/hpsa.py
index 648e1fc..3cf93ce 100644
--- a/plugin/hpsa/hpsa.py
+++ b/plugin/hpsa/hpsa.py
@@ -576,15 +576,22 @@ def _hp_ld_to_lsm_vol(hp_ld, pool_id, sys_id, ctrl_num, array_num,
         # No document or command output indicate block size
         # of volume. So we try to read from linux kernel, if failed
         # try 512 and roughly calculate the sector count.
-        device = Device.from_device_file(_CONTEXT, hp_ld['Disk Name'])
-        vol_name = "%s: /dev/%s" % (hp_ld_name, device.sys_name)
-        attributes = device.attributes
-        try:
-            block_size = attributes.asint("queue/logical_block_size")
-            num_of_blocks = attributes.asint("size")
-        except (KeyError, UnicodeDecodeError, ValueError):
-            block_size = 512
-            num_of_blocks = int(_hp_size_to_lsm(hp_ld['Size']) / block_size)
+        block_size = 512
+        num_of_blocks = int(_hp_size_to_lsm(hp_ld['Size']) / block_size)
+        vol_name = hp_ld_name
+
+        if len(vpd83) > 0:
+            blk_paths = LocalDisk.vpd83_search(vpd83)
+            if len(blk_paths) > 0:
+                blk_path = blk_paths[0]
+                vol_name += ": %s" % " ".join(blk_paths)
+                device = Device.from_device_file(_CONTEXT, blk_path)
+                attributes = device.attributes
+                try:
+                    block_size = attributes.asint("queue/logical_block_size")
+                    num_of_blocks = attributes.asint("size")
+                except (KeyError, UnicodeDecodeError, ValueError):
+                    pass
 
         if 'Failed' in hp_ld['Status']:
             admin_status = Volume.ADMIN_STATE_DISABLED
-- 
2.9.0

```

Comment (joehandzik) [inline plugin/hpsa/hpsa.py]: @cvedel Any way for you to give this a try instead? I think it's better than the PR I submitted, and this is likely the fix we will go with.

**Comment by cathay4t** [comment]
@joehandzik I reproduced this problem via:
```bash
lsmcli vrc --name test --raid-type raid0 --disk 0TV575LJ
ID                               | Name                        | SCSI VPD 0x83                    | Size         | Disabled | Pool ID                | System ID      | Disk Paths
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
600508b1001c11eb2f951679b37d104e | Logical Drive: 2: /dev/sdap | 600508b1001c11eb2f951679b37d104e | 299966445568 | No       | PDNNF0ARH6W07M:Array:B | PDNNF0ARH6W07M | /dev/sdap 

echo 1 | sudo tee /sys/block/sdap/device/delete

lsmcli lv
PLUGIN_BUG(2): Expected key missing from SmartArray hpssacli output:'Disk Name'
```

My patch fixed this problem:
```bash
lsmcli lv
ID                               | Name                       | SCSI VPD 0x83                    | Size         | Disabled | Pool ID                | System ID      | Disk Paths
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
600508b1001c79ade5178f0626caaa9c | Logical Drive: 1: /dev/sda | 600508b1001c79ade5178f0626caaa9c | 299966445568 | No       | PDNNF0ARH6W07M:Array:A | PDNNF0ARH6W07M | /dev/sda  
600508b1001c11eb2f951679b37d104e | Logical Drive: 2           | 600508b1001c11eb2f951679b37d104e | 300003465216 | No       | PDNNF0ARH6W07M:Array:B | PDNNF0ARH6W07M |
```

**Comment by cathay4t** [comment]
Closing this PR as commit 553770fb002ef24429a86f816986d74297ef69f2 fixed this issue.

**Comment by joehandzik** [comment]
@cathay4t sounds good, thanks!

**Pull request closed without merging**